{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1gs4nShMY8OVf9TEYxx6g6zqQ3vuEbqyL","timestamp":1724322441075}]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["## Exception Handling\n"],"metadata":{"id":"ajVfqio_GinI"}},{"cell_type":"markdown","source":["###`Q-1`: You are given a function definition. There might be several issues on execution of this function. You are asked to do exception handling for diffrent errors that this function goes in to `without altering this function`. And print error text.\n","\n"],"metadata":{"id":"0qnQkJtnGozz"}},{"cell_type":"markdown","source":["Function parameters `l -> list, s -> could be anything`\n","\n","```\n","def function(l: list, s, **args):\n","    last_element = l[-1]\n","    \n","    l[int(s)]=10\n","    any_element = l[int(s)+10]\n","    l[s]=10\n","    \n","    res = sum(l)\n","    \n","    p = args['p']\n","    # print(p)\n","    return res/last_element * p + any_element\n","\n","```\n","Check for different function calls:-\n","\n","```\n","function([1,2,1], 12)\n","function([1,2,1]*9, '1-2')\n","function([1,'2',1]*9, 12)\n","function([1,'2',1]*9, 12)\n","function([1,2,0]*9, 12  )\n","function([1,2,1]*9, 12, p=None)\n","function([1,2,0]*9, 12, p=10)\n","```"],"metadata":{"id":"2no_EB2fGyGq"}},{"cell_type":"code","source":["def function(l: list, s, **args):\n","    last_element = l[-1]\n","    l[int(s)]=10\n","    any_element = l[int(s)+10]\n","    l[s]=10\n","\n","    res = sum(l)\n","\n","    p = args['p']\n","    # print(p)\n","    return res/last_element * p + any_element\n","\n","\n","try:\n","  function([1,2,1], 12)\n","except IndexError as e:\n","  print(type(e))\n","  print(e)\n","\n","try:\n","  function([1,2,1]*9, '1-2')\n","except ValueError as v:\n","  print(type(v))\n","  print(v)\n","\n","try:\n","  function([1,'2',1]*9, 12)\n","except TypeError as v:\n","  print(type(v))\n","  print(v)\n","\n","try:\n","  function([1,2,0]*9, 12  )\n","except KeyError as v:\n","  print(type(v))\n","  print(v)\n","\n","try:\n","  function([1,2,0]*9,12, p = None  )\n","except ZeroDivisionError as v:\n","  print(type(v))\n","  print(v)\n","\n","try:\n","  function([1,2,0]*9, 12, p=10)\n","except ZeroDivisionError as v:\n","  print(type(v))\n","  print(v)\n","try:\n","  res = function([1,2,1]*9, 12, p = 10)\n","except Exception as e:\n","  print(e)\n","else:\n","  print(\"Result:\",res)\n","finally:\n","  print(\"thank you\")\n","\n"],"metadata":{"id":"VoaL4_9eGjFU","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1724326549630,"user_tz":-330,"elapsed":715,"user":{"displayName":"Sunil Kumar","userId":"07008552861197289083"}},"outputId":"b0a63007-3c1e-4784-b84f-503df349e3fe"},"execution_count":35,"outputs":[{"output_type":"stream","name":"stdout","text":["<class 'IndexError'>\n","list assignment index out of range\n","<class 'ValueError'>\n","invalid literal for int() with base 10: '1-2'\n","<class 'TypeError'>\n","unsupported operand type(s) for +: 'int' and 'str'\n","<class 'KeyError'>\n","'p'\n","<class 'ZeroDivisionError'>\n","division by zero\n","<class 'ZeroDivisionError'>\n","division by zero\n","Result: 452.0\n","thank you\n"]}]},{"cell_type":"markdown","source":["###`Q-2:` You are given a code snippet. There might be several issues on execution of this code. You are asked to do exception handling for diffrent errors, condition is what ever happens we need to execute last line printing correct result of `sum of elements`.\n","\n","List have elemnts as any no of  `key-pair dict with key as list index and value as any integer`, `integers` and `numeric-strings`. There is always only one element in the dict.\n","\n","\n","```\n","l = [{0:2},2,3,4,'5', {5:10}]\n","# For calculating sum of above list\n","s=0\n","for i in range(len(l)):\n","    #You can Edit code from here\n","    s += l[i].get(i)\n","    s += l[i]\n","    s += int(l[i])\n","\n","\n","print(s)\n","```"],"metadata":{"id":"HuLMy2fVG9sL"}},{"cell_type":"code","source":["l = [{0:2},2,3,4,'5', {5:10}]\n","# For calculating sum of above list\n","s=0\n","for i in range(len(l)):\n","    #You can Edit code from here\n","    try:\n","      s += l[i]\n","    except TypeError:\n","      try:\n","        s += l[i].get(i)\n","      except AttributeError:\n","        s += int(l[i])\n","\n","\n","\n","print(s)"],"metadata":{"id":"pEP8JU0qHAf8","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1724328193814,"user_tz":-330,"elapsed":500,"user":{"displayName":"Sunil Kumar","userId":"07008552861197289083"}},"outputId":"a83d0943-3cc2-4c9f-9bd9-b849b34d27ac"},"execution_count":39,"outputs":[{"output_type":"stream","name":"stdout","text":["26\n"]}]},{"cell_type":"markdown","source":["### `Q-3:`: File Handling with Exception handling\n","\n","Write a program that opens a text file and write data to it as \"Hello, Good Morning!!!\". Handle exceptions that can be generated during the I/O operations. Do not show the success message on the main exception handling block (write inside the else block)."],"metadata":{"id":"ohE2fDTXHIan"}},{"cell_type":"code","source":["\n","try:\n","  with open(\"sample.txt\",\"w\") as f:\n","    f.write(\"Hello, Good Morning!!!\")\n","except IOError:\n","  print(\"File not Created\")\n","else:\n","  print(\"File created sucessfully\")\n","  with open(\"sample.txt\",'r') as r:\n","    print(r.read())\n"],"metadata":{"id":"iOeiWrcIHPlY","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1724328829358,"user_tz":-330,"elapsed":523,"user":{"displayName":"Sunil Kumar","userId":"07008552861197289083"}},"outputId":"df2e7f70-e784-4acc-d8d8-fb9ef8bd5ea7"},"execution_count":45,"outputs":[{"output_type":"stream","name":"stdout","text":["File created sucessfully\n","Hello, Good Morning!!!\n"]}]},{"cell_type":"markdown","source":["### `Q-4`: Number game program.\n","\n","Write a number game program. Ask the user to enter a number. If the number is greater than number to be guessed, raise a **ValueTooLarge** exception. If the value is smaller the number to be guessed the, raise a **ValueTooSmall** exception and prompt the user to enter again. Quit the program only when the user enters the correct number. Also raise **GuessError** if user guess a number less than 1."],"metadata":{"id":"ru8g8-_bHTUf"}},{"cell_type":"code","source":["import random"],"metadata":{"id":"QftlN5LJHZSX"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### `Q-5:` Cast vote\n","\n","Write a program that validate name and age as entered by the user to determine whether the person can cast vote or not. To handle the age, create **InvalidAge** exception and for name, create **InvalidName** exception. The name will be invalid when the string will be empty or name has only one word.\n","\n","Example 1:\n","\n","Input:\n","\n","```bash\n","Enter the name:               goransh singh\n","Enter the age: 25\n","```\n","\n","Output:\n","\n","```bash\n","Goransh Singh  Congratulation !!! You can vote.\n","```"],"metadata":{"id":"ZsGI_40fHcsd"}},{"cell_type":"code","source":["# Write code here"],"metadata":{"id":"OHL3UiS4Hi3U"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### `Q-6`: Write a python function which infinitely prints natural numbers in a single line. Raise the **StopIteration** exception after displaying first 20 numnbers to exit from the program."],"metadata":{"id":"SCvs0QjgHlOG"}},{"cell_type":"code","source":["# Write code here"],"metadata":{"id":"Lc5ilNgNHoSx"},"execution_count":null,"outputs":[]}]}